### PR TITLE
Fix invalid UPDATE stmt when none of the columns change

### PR DESCRIPTION
**Pull request by arajkumar**
When all column values in the SET clause are equal to those in the WHERE clause, we remove all columns from the SET clause. This results in an invalid UPDATE statement like the one shown below:
```
UPDATE table SET WHERE "id" = 1;
```

Usually, the above could happen when REPLICA IDENTITY is set to FULL, and the UPDATE statement executed with the same values as the old ones.
For e.g.
```
UPDATE table SET "id" = 1 WHERE "id" = 1;
```

**Solution**: Skip the UPDATE when all column values in SET clause are equal to the WHERE clause values.

Fixes https://github.com/dimitri/pgcopydb/issues/750

### CODE REVIEW

**Comment by arajkumar** [comment]
Hmmm, I think the alternative approach is the right one.

**Comment by dimitri** [comment]
Do we need to worry about triggers on the target side? UPDATE statement that look like they change nothing can be used to execute a trigger... on the replica system though, this would have to be a specifically crafted trigger that still is executed with session_replication_role set to replica.

**Comment by arajkumar** [comment]
@dimitri In this PR, I'm skipping the UPDATE stmt when all columns in SET clause are duplicates of WHERE clause values. 

Once we fix https://github.com/dimitri/pgcopydb/issues/844, we shall also remove the duplicate removal logic.

File: src/bin/pgcopydb/ld_transform.c [hunk 1]
```diff
@@ -2227,12 +2227,15 @@ stream_write_update(FILE *out, LogicalMessageUpdate *update)
 						  update->table.relname);
 		int pos = 0;
 
+		int skippedColumns = 0;
```
Comment (dimitri) [inline src/bin/pgcopydb/ld_transform.c:2230]: Can re rename to `skipColumnCount` or something? at the moment this reads more like a boolean value to me.

File: src/bin/pgcopydb/ld_transform.c [hunk 2]
```diff
@@ -2360,20 +2367,44 @@ stream_write_update(FILE *out, LogicalMessageUpdate *update)
...
+		 * Skip the UPDATE statement in such cases.
+		 */
+		bool skipUpdate = (skippedColumns == new->attributes.count);
 
-		FFORMAT(out, "PREPARE %x AS %s;\n", hash, buf->data);
+		if (skipUpdate)
+		{
+			log_warn("Skipping UPDATE statement as all columns are "
+					 "the same as the old");
+		}
```
Comment (dimitri) [inline src/bin/pgcopydb/ld_transform.c:2390]: If we would `return true;` from here, we wouldn't need an else branch at all. What do you think?

**Review by dimitri — COMMENTED** [review]
Finally had time to review. Thanks for the great work again!

I was initially confused about what we already do in pgcopydb before your PR and what's added in your PR. I now got clarity on that, and appreciate your changes. Apart from the nitpicking in the review, I think we should add some mention of what we're doing here in the docs.

Maybe between https://pgcopydb.readthedocs.io/en/latest/ref/pgcopydb_follow.html#replica-identity-and-lack-of-primary-keys and https://pgcopydb.readthedocs.io/en/latest/ref/pgcopydb_follow.html#logical-decoding-pre-fetching in a new section titled “Statement Rewriting” or something.

**Comment by arajkumar** [comment]
@dimitri Thanks for reviewing this PR.

>  I think we should add some mention of what we're doing here in the docs.

I don't think we should add to the doc, This is just a big fix to avoid generation of invalid SQL UPDATE statement. Also, I believe we should just implement skipping identity columns being included as part of UPDATE's SET clause(https://github.com/dimitri/pgcopydb/issues/844) and remove the deduplication logic all together. WDYT?

**Review by dimitri — APPROVED** [review]

**Comment by dimitri** [comment]
> I don't think we should add to the doc, This is just a big fix to avoid generation of invalid SQL UPDATE statement.

As a change of behavior, I think you're right, no need to document the fix. That said, I think our overall approach should be documented. Saying that makes me realize it's not fair to add that burden to your PR (now merged), I will think more about what I think is needed and take care of it.

> Also, I believe we should just implement skipping identity columns being included as part of UPDATE's SET clause(#844) and remove the deduplication logic all together. WDYT?

First part, agreed. Second part (remove the deduplication logic), I'm not sure I understand it. Can you provide a couple examples showing what we do, why it's wrong, and what you propose we should do instead?

**Pull request merged**